### PR TITLE
Also discard 'caches' and 'backfill' stream POSITIONS

### PR DESCRIPTION
**Pull request by erikjohnston**
Follow on from #16640

### CODE REVIEW

File: synapse/replication/tcp/streams/_base.py [hunk 1]
```diff
@@ -305,6 +305,14 @@ def minimal_local_current_token(self) -> Token:
         # which means we need to negate it.
         return -self.store._backfill_id_gen.get_minimal_local_current_token()
 
+    def can_discard_position(
+        self, instance_name: str, prev_token: int, new_token: int
+    ) -> bool:
+        # Caches streams can't go backwards, so we know we can ignore any
+        # positions where the tokens are from before the current token.
```
Comment (clokep) [inline synapse/replication/tcp/streams/_base.py:312]: This isn't the caches stream?

File: changelog.d/16655.misc
Comment (clokep) [inline changelog.d/16655.misc:1]: Do we no longer need #16557 (is this a more general solution)?

File: synapse/replication/tcp/streams/_base.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/replication/tcp/streams/_base.py:312]: Oh, I forgot I did the backfill here too

File: synapse/replication/tcp/streams/_base.py [hunk 2]
```diff
@@ -305,6 +305,14 @@ def minimal_local_current_token(self) -> Token:
         # which means we need to negate it.
         return -self.store._backfill_id_gen.get_minimal_local_current_token()
 
+    def can_discard_position(
+        self, instance_name: str, prev_token: int, new_token: int
+    ) -> bool:
+        # Caches streams can't go backwards, so we know we can ignore any
```
Comment (erikjohnston) [inline synapse/replication/tcp/streams/_base.py:311]:
```suggestion
        # Backfill stream can't go backwards, so we know we can ignore any
```

File: changelog.d/16655.misc
Comment (erikjohnston) [reply changelog.d/16655.misc:1]: Not really, that PR reduces the amount *sent*, this is reducing the processing of what we receive (but its still non-zero)

Comment (clokep) [reply changelog.d/16655.misc:1]: I thought #16561 reduced the amount sent? I guess #16557 does so too though.

**Review by clokep — APPROVED** [review]
It looks like this the ones that we don't apply this to are the `PresenceFederationStream` and `TypingStream` (the two direct sub-classes of `Stream` without the additional logic).

**Pull request merged**